ot/gromov/_utils.py

```diff
@@ -522,3 +524,254 @@ def h2(b):
+    """
+    Calculate the KL divergence between two product measures:
+    KL(mu \otimes nu, alpha \otimes beta) =
+    m_mu * KL(nu, beta) + m_nu * KL(mu, alpha) +
+    (m_mu - m_alpha) * (m_nu - m_beta)
+
+    Parameters
+    ----------
+    mu: vector or matrix
+    nu: vector or matrix
+    alpha: vector or matrix with the same size as mu
+    beta: vector or matrix with the same size as nu
+
+    Returns
+    ----------
+    KL divergence between two product measures
+    """
+
+    def kl(p, q, nx=None):
+        return approx_kl(p, q, nx=nx) - nx.sum(p) + nx.sum(q)
+
+    if nx is None:
+        mu, nu, alpha, beta = list_to_array(mu, nu, alpha, beta)
+        nx = get_backend(mu, nu, alpha, beta)
+
+    m_mu, m_nu = nx.sum(mu), nx.sum(nu)
+    m_alpha, m_beta = nx.sum(alpha), nx.sum(beta)
+    const = (m_mu - m_alpha) * (m_nu - m_beta)
+    res = m_nu * kl(mu, alpha, nx) + m_mu * kl(nu, beta, nx) + const
+
+    return res
+
+
+def product_l2(mu, nu, alpha, beta, nx=None):
+    """
+    norm = ||mu \otimes nu - alpha \otimes beta ||^2
+    = ||a||^2 ||b||^2 + ||mu||^2 ||nu||^2 - 2 < alpha, mu > < beta, nu >.
+    L2(mu \otimes nu, alpha \otimes beta) = norm / 2.
+    """
+
+    if nx is None:
+        mu, nu, alpha, beta = list_to_array(mu, nu, alpha, beta)
+        nx = get_backend(mu, nu, alpha, beta)
+
+    norm = nx.sum(alpha**2) * nx.sum(beta**2) \
+        - 2 * nx.sum(alpha * mu) * nx.sum(beta * nu) \
+        + nx.sum(mu**2) * nx.sum(nu**2)
+
+    return norm / 2
+
+
+def product_div(mu, nu, alpha, beta, divergence, nx=None):
+    if nx is None:
+        mu, nu, alpha, beta = list_to_array(mu, nu, alpha, beta)
+        nx = get_backend(mu, nu, alpha, beta)
+
+    if divergence == "kl":
+        return product_kl(mu, nu, alpha, beta, nx)
+    elif divergence == "l2":
+        return product_l2(mu, nu, alpha, beta, nx)
+
+
+# Support functions for BCD schemes
+def local_cost(data, pi, tuple_p, hyperparams, divergence, reg_type, nx=None):
+    """
+    Calculate cost matrix of the UOT subroutine
+    """
+
+    X_sqr, Y_sqr, X, Y, M = data
+    rho_x, rho_y, eps = hyperparams
+    a, b = tuple_p
+
+    if nx is None:
+        X, Y, a, b = list_to_array(X, Y, a, b)
+        nx = get_backend(X, Y, a, b)
+
+    pi1, pi2 = nx.sum(pi, 1), nx.sum(pi, 0)
+    A, B = nx.dot(X_sqr, pi1), nx.dot(Y_sqr, pi2)
+    uot_cost = A[:, None] + B[None, :] - 2 * nx.dot(nx.dot(X, pi), Y.T)
+    if M is not None:
+        uot_cost = uot_cost + M
+
+    if divergence == "kl":
+        if rho_x != float("inf") and rho_x != 0:
+            uot_cost = uot_cost + rho_x * approx_kl(pi1, a, nx)
+        if rho_y != float("inf") and rho_y != 0:
+            uot_cost = uot_cost + rho_y * approx_kl(pi2, b, nx)
+        if reg_type == "joint" and eps > 0:
+            uot_cost = uot_cost + eps * approx_shortcut_kl(pi, pi1, pi2, a, b, nx)
+
+    return uot_cost
+
+
+def total_cost(M_linear, data, tuple_pxy_samp, tuple_pxy_feat,
```

Comment:
improve name too + doc. Plus potential simplification of the function thanks to `local_cost` function.